### PR TITLE
Decouple engine and client

### PR DESCRIPTION
**Pull request by rtibbles**
Separate out the engine that runs the scheduler and the workers and the client that allows jobs to be scheduled and canceled.

### CODE REVIEW

**Comment by indirectlylit** [comment]
One thought for the longer-term (not necessarily this PR):

Would be nice to align the iceqube API with [redis queue](http://python-rq.org/docs/) and [django rq](https://github.com/rq/django-rq). This would enable us to more easily swap in a higher-performance system when it's available, e.g. under `kolibri-server`.

In particular I think our 'client' is a 'queue' in RQ:

```
queue = Queue()
job = q.enqueue(func, args)
```

and our 'engine' is a forked 'worker' process in RQ:

```
w = Worker(queue)
w.work()
```

File: src/iceqube/messaging/backends/inmem.py [hunk 1]
```diff
@@ -62,21 +16,6 @@ def send(self, mailbox, message, *args, **kwargs):
         INMEM_SUPER_MAILBOX[mailbox]._put(message)
         logger.debug("SEND MAILBOX: {} MSG: {}".format(mailbox, message))
 
-    def wait(self, mailbox, timeout=None):
```
Comment (aronasorman) [inline src/iceqube/messaging/backends/inmem.py]: Hmm where did you put the `wait` method? It seems like this and the abstract base class stub was removed as well.

Comment (rtibbles) [reply src/iceqube/messaging/backends/inmem.py]: Yep, it wasn't used anywhere, so I got rid of it.

Comment (aronasorman) [reply src/iceqube/messaging/backends/inmem.py]: Ah, gotcha.

**Comment by rtibbles** [comment]
Also now fixes #38 - happy to switch naming convention if required.

**Comment by rtibbles** [comment]
This now also adds a `NoConfig` engine and client that tries to run with process based workers and a socket based messaging backend, but falls back to threaded workers with a shared memory messaging backend if the environment does not support it.

**Comment by rtibbles** [comment]
Ran a benchmark of JSON (which I can pretty much reduce my wire protocol to) against `msgpack` in Pure Python fallback, and `umsgpack` another Pure Python implementation of the msgpack wire protocol:
```
array size: 0
json with dict 3,688.19952011ns/pack 4,702.42977142ns/unpack (82 bytes)
json with tuple 2,973.549366ns/pack 2,621.21915817ns/unpack (42 bytes)
umsgpack with dict 24,693.4580803ns/pack 22,692.4395561ns/unpack (66 bytes)
umsgpack with tuple 11,848.988533ns/pack 8,608.93964767ns/unpack (34 bytes)
msgpack with dict 20,518.2695389ns/pack 26,719.1314697ns/unpack (60 bytes)
msgpack with tuple 13,684.8306656ns/pack 16,025.2904892ns/unpack (32 bytes)
array size: 10
json with dict 4,526.88217163ns/pack 5,500.19979477ns/unpack (110 bytes)
json with tuple 4,190.02056122ns/pack 3,796.11968994ns/unpack (70 bytes)
umsgpack with dict 32,728.729248ns/pack 32,426.6505241ns/unpack (76 bytes)
umsgpack with tuple 23,194.4513321ns/pack 17,113.0180359ns/unpack (44 bytes)
msgpack with dict 31,221.1799622ns/pack 38,541.1691666ns/unpack (70 bytes)
msgpack with tuple 21,561.5200996ns/pack 30,094.8095322ns/unpack (42 bytes)
array size: 100
json with dict 12,891.1495209ns/pack 14,519.200325ns/unpack (470 bytes)
json with tuple 13,649.1298676ns/pack 10,184.8101616ns/unpack (430 bytes)
umsgpack with dict 123,862.068653ns/pack 109,965.941906ns/unpack (168 bytes)
umsgpack with tuple 117,196.419239ns/pack 100,736.33194ns/unpack (136 bytes)
msgpack with dict 112,326.660156ns/pack 139,469.320774ns/unpack (162 bytes)
msgpack with tuple 103,888.64994ns/pack 129,492.080212ns/unpack (134 bytes)
```

JSON is a clear winner here - so I will update the existing implementation based on these benchmarks.

**Comment by rtibbles** [comment]
This no longer uses an inefficient and non-standard wire protocol.

**Comment by rtibbles** [comment]
Possibly, although I added the `usedevelop` change because without it, I couldn't run tests at all.

File: src/iceqube/worker/backends/inmem.py [hunk 2]
```diff
@@ -83,6 +76,77 @@ def handle_finished_future(self, future):
...
+        If fewer workers are running than there are available workers, start a new job!
+        Returns: None
+        """
+        try:
+            while len(self.future_job_mapping.keys()) < self.num_workers:
+                self.start_next_job()
+        except Empty:
+            logger.debug("No jobs to start.")
+        cancelling_jobs = [job.job_id for job in self.storage_backend.get_canceling_jobs()]
+        if cancelling_jobs:
```
[the remark this answers is not in the document]
Comment (rtibbles) [reply src/iceqube/worker/backends/inmem.py]: Yeah, and I made that change locally already as I still need to trigger events for the canceling jobs that are *not* currently being run.

**Comment by rtibbles** [comment]
Ignore the branch check... I accidentally pushed to upstream, then deleted it.

**Comment by rtibbles** [comment]
Did one last refactor to flatten and simplify the file structure and remove unnecessary functionality.

**Comment by rtibbles** [comment]
@bjester does the latest commit do anything to satisfy the concerns you expressed in #40? (I simplified things so that a factory pattern would be unnecessary, and tried to remove the last of the unneeded abstraction).

**Comment by bjester** [comment]
@rtibbles Looks better now that some of it's cleaned up. I think I would label #40 as an enhancement, and since this is a standalone framework that others could benefit from, I still think incorporating some of those methodologies could be very beneficial for extensibility and testing.

File: src/iceqube/queue.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    @property
+    def jobs(self):
+        """
+        Return all the jobs scheduled, queued, running, failed or completed.
+        Returns: A list of all jobs.
+
+        """
+        return self.storage.get_all_jobs()
+
+    def enqueue(self, func, *args, **kwargs):
```
Comment (aronasorman) [inline src/iceqube/queue.py]: Is this the official API for iceqube now?

File: tests/test_classes.py [hunk 4]
```diff
@@ -213,30 +204,30 @@ def test_stringify_func_is_importable(self):
...
+        job_id = inmem_queue.enqueue(
             cancelable_job,
-            is_running_event=is_running_event,
-            is_not_canceled_event=is_not_canceled_event,
             cancellable=True
         )
 
-        is_running_event.wait(1.0)
+        job = inmem_queue.fetch_job(job_id)
+        while job.state != State.RUNNING:
```
Comment (aronasorman) [inline tests/test_classes.py]: We should add a timeout here, or a catch so we don't get into an infinite loop.

File: tests/test_classes.py [hunk 5]
```diff
@@ -213,30 +204,30 @@ def test_stringify_func_is_importable(self):
...
-        inmem_client.cancel(job_id)
+        inmem_queue.cancel(job_id)
         # And check the job state to make sure it's marked as cancelling
-        job = inmem_client.status(job_id)
+        job = inmem_queue.fetch_job(job_id)
         assert job.state == State.CANCELING
 
-        # Let's wait for another job state change...
-        job = inmem_client.wait_for_completion(job_id, timeout=2.0)
+        while job.state != State.CANCELED:
```
Comment (aronasorman) [inline tests/test_classes.py]: Same here too, we wanna have a limit to the number of times we loop.

File: src/iceqube/queue.py (hunk 3 above)
Comment (rtibbles) [reply src/iceqube/queue.py]: Yeah, this is parallel to RQ's API.

File: tests/test_classes.py (hunk 4 above)
Comment (rtibbles) [reply tests/test_classes.py]: Fair enough.

**Comment by rtibbles** [comment]
Yeah, no more messaging - relying on the DB entirely as the single source of truth, and flattening and removing abstractions.

File: tox.ini [hunk 6]
```diff
@@ -11,7 +11,7 @@ python =
   pypy: pypypy
 
 [testenv]
-usedevelop = false
+usedevelop = True
```
Comment (benjaoming) [inline tox.ini]: This change is risky, if something makes the tests depend on development code, what guarantees that the distributed code then works?

Opened https://github.com/learningequality/iceqube/issues/42

Comment (rtibbles) [reply tox.ini]: It didn't work without it, and it's what we use in Kolibri.

Would be good to know what we should do to make our tests run against production packages.

Comment (benjaoming) [reply tox.ini]: https://github.com/learningequality/kolibri/issues/3310 :wink:

**Pull request merged**